main/main.go

```diff
@@ -199,11 +212,85 @@ func usage() {
 	flag.PrintDefaults()
 }
 
+type diffToolStateType int
+
+const (
+	finStateInitial diffToolStateType = iota
+	dcpDriving      diffToolStateType = iota
+	finStateFinal   diffToolStateType = iota
+)
+
+type difftoolState struct {
+	state diffToolStateType
+	mtx   sync.Mutex
+}
+
+type xdcrDiffTool struct {
+	utils              xdcrUtils.UtilsIface
+	metadataSvc        service_def.MetadataSvc
+	remoteClusterSvc   service_def.RemoteClusterSvc
+	replicationSpecSvc service_def.ReplicationSpecSvc
+	logger             *xdcrLog.CommonLogger
+
+	specifiedRef  *metadata.RemoteClusterReference
+	specifiedSpec *metadata.ReplicationSpecification
+	filter        xdcrParts.FilterIface
+
+	sourceDcpDriver *dcp.DcpDriver
+	targetDcpDriver *dcp.DcpDriver
+
+	curState difftoolState
+}
+
+func NewDiffTool() *xdcrDiffTool {
+
+	difftool := &xdcrDiffTool{
+		utils: xdcrUtils.NewUtilities(),
+	}
+	difftool.metadataSvc, _ = metadata_svc.NewMetaKVMetadataSvc(nil, difftool.utils)
+
+	uiLogSvcMock := &service_def_mock.UILogSvc{}
+	xdcrTopologyMock := &service_def_mock.XDCRCompTopologySvc{}
+	clusterInfoSvcMock := &service_def_mock.ClusterInfoSvc{}
+
+	difftool.logger = xdcrLog.NewLogger("xdcrDiffTool", nil)
+
+	difftool.remoteClusterSvc, _ = metadata_svc.NewRemoteClusterService(uiLogSvcMock, difftool.metadataSvc, xdcrTopologyMock,
+		clusterInfoSvcMock, xdcrLog.DefaultLoggerContext, difftool.utils)
+
+	difftool.replicationSpecSvc, _ = metadata_svc.NewReplicationSpecService(uiLogSvcMock, difftool.remoteClusterSvc,
+		difftool.metadataSvc, xdcrTopologyMock, clusterInfoSvcMock,
+		nil, difftool.utils)
+
+	// Capture any Ctrl-C for continuing to next steps or cleanup
+	go difftool.monitorInterruptSignal()
+
+	return difftool
+}
+
+func maybeSetEnv(key, value string) {
+	if os.Getenv(key) != "" {
+		return
+	}
+	os.Setenv(key, value)
+}
+
 func main() {
 	argParse()
 
+	difftool := NewDiffTool()
+
+	if len(options.remoteClusterName) > 0 {
```

Comment:
this part needs some more thoughts.
it seems that remoteClusterName is supposed to be not empty when filtering is involved, and empty otherwise. this is not intuitive or flexible.
1. when not having filtering, user may want to use clusterName to avoid typing in target cluster info
2. when having filtering, user may want to connect to target cluster in a different way from that specified in the remoteClusterRef (e.g., clusterRef has SSL enabled, and tool does not need SSL).

Reply:
1. UUID check is now added.
2. With current design, users can use remote cluster name regardless of filtering or not.
3. Fixed script to pass in target username and pw, and also add them into spec once retrieved.